lib/model/receiveonlyfolder.go

```diff
@@ -0,0 +1,158 @@
+
+	return f
+}
+
+func (f *receiveOnlyFolder) String() string {
+	return fmt.Sprintf("receiveOnlyFolder/%s@%p", f.folderID, f)
+}
+
+// validateAndUpdateLocalChanges reverts all local changes
+func (f *receiveOnlyFolder) validateAndUpdateLocalChanges(fs []protocol.FileInfo) []protocol.FileInfo {
+	fileDeletions := []protocol.FileInfo{}
+	dirDeletions := []protocol.FileInfo{}
+
+	for i, file := range fs {
+		if strings.Contains(file.Name, ".sync-conflict-") {
+			// this is a conflict copy, let's move on to the next file
+			continue
+		}
+
+		objType := "file"
+		action := "modified"
+		correctiveAction := "resync"
+
+		if file.IsDirectory() {
+			objType = "dir"
+		}
+
+		if file.IsDeleted() {
+			action = "deleted"
+		}
+
+		if len(file.Version.Counters) == 1 && file.Version.Counters[0].Value == 1 {
+			// A file, directory or symlink was added, which we'll have to remove again
+			action = "added"
+			if f.DeleteLocalChanges {
+				correctiveAction = "deleted"
+				if file.IsDirectory() {
+					dirDeletions = append(dirDeletions, file)
+				} else {
+					fileDeletions = append(fileDeletions, file)
+				}
+			} else {
+				correctiveAction = "none"
+			}
+		}
+		// let's update the record to reflec that this is invalid and should be pulled again if possible
+		fs[i].Deleted = false
+		fs[i].Invalid = true
+		fs[i].Version = protocol.Vector{}
+
+		// we better tell the user on the UI and in the log that we had to take corrective actions
+		l.Infoln("Rejecting local change on folder", f.Description(), objType, file.Name, "was", action, "corrective action:", correctiveAction)
+
+		// Fire the LocalChangeRejected event to notify listeners about rejected local changes.
+		events.Default.Log(events.LocalChangeRejected, map[string]string{
+			"folder": f.ID,
+			"item":   file.Name,
+			"type":   objType,
+			"action": correctiveAction,
+		})
+	}
+
+	// delete all the files first, so versioning and conflict managed gets applied
+	for _, file := range fileDeletions {
+		l.Debugln("Deleting file", file.Name)
+		f.deleteRejectedFile(file, f.versioner)
+	}
+
+	// now get rid of those pesky directories that were created
+	for i := range dirDeletions {
+		dir := dirDeletions[len(dirDeletions)-i-1]
+		l.Debugln("Deleting dir", dir.Name)
+		f.deleteRejectedDir(dir)
+	}
+
+	// update the database
+	f.model.updateLocals(f.ID, fs)
+
+	// trigger a pull
+	f.IndexUpdated()
+
+	// return the update list of files
+	return fs
+}
+
+// deleteRejectedDir attempts to delete the given directory
+func (f *receiveOnlyFolder) deleteRejectedDir(file protocol.FileInfo) {
+	err := f.folder.deleteDir(f.Path(), file, nil)
+	if err != nil && !os.IsNotExist(err) {
+		l.Infof("deleteRejectedDir (folder %q, file %q): delete: %v", f, file.Name, err)
+	}
+}
+
+// deleteRejectedFile attempts to delete the given file
+func (f *receiveOnlyFolder) deleteRejectedFile(file protocol.FileInfo, ver versioner.Versioner) {
+	err := f.folder.deleteFile(f.Path(), file, ver, f.MaxConflicts)
```

Comment:
Not sure if it's possible to change sendReceiveFolder to embed folder fully, then these could just be f.deleteFile, if not, then it's fine as it is, though yeah, it's not for this pull request either.